Add documentation pages about preloading and FRONT_COMMERCE_FAST (#104)
front-commerce/developers.front-commerce.com: source/docs/advanced/performance/preloading-routes.md

---
id: preloading-routes
title: Preloading Routes' data
---

> 

Reviewer: I think we should add this here for the whole page for now. At some point we could remove all those notices when we will be confident that all projects use 1.0 for instance (and later, after EOL).

```suggestion

<blockquote class="feature--new">
  _This feature has been added in version `1.0.0-rc.0`_
</blockquote>

```

When a website is slow, the first thing that comes to mind is to improve raw metrics like the Speed Index. And while this is definitly useful, we shouldn't overlook the [_perceived_ performances](https://blog.teamtreehouse.com/perceived-performance). The goal is not to improve the response time, but to better manage the time to make the user _think_ that the website is fast.

> One way to do this is to anticipate the user's need and start the long running tasks before they actually need it. For instance, when a user is navigating on a catalog, on desktop, they will hover the product before clicking on it. Thus, we can trigger the product's load as soon as they hover the product, instead of waiting for the click. This can seem like small gains, but this definitly improves how fast the user _thinks_ your site is.

Reviewer: ```suggestion
One way to do this is to anticipate the user's need and start the long running tasks before they actually need it. For instance, when a user is navigating on a catalog, on desktop, they will hover the product before clicking on it. Thus, we can start loading the product page as soon as the user hover the product element, instead of waiting for the click. This can seem like a micro-optimization, but it definitely improves how fast the user _thinks_ your site is.
```

## How to preload data?

The main idea is to launch Apollo requests twice:

- when you think that the user will need the data
- when you are displaying the data

> By doing this, you will first warm up Apollo's cache and this cache will be reused when you will display your data. But this can be tedious. This is why in Front-Commerce, we have added a <abbr title="Higher Order Component">HOC</abbr> that will do this: `web/core/apollo/graphqlWithPreload` (added in version 1.0.0-rc.0).

Reviewer: ```suggestion
By doing this, you will first warm up Apollo's cache and this cache will be reused when you will display your data. But this can be tedious. This is why in Front-Commerce, we have added a <abbr title="Higher Order Component">HOC</abbr> that will do this: `web/core/apollo/graphqlWithPreload`.
```

For instance, let's say that we want to preload the product with SKU `VD12` when you click on a button. This means that you will write something like:

```js
const PreloadProductButton = () => {
  return (
    <button
      onClick={() => {
        EnhancedProduct.preload({ sku: "VD12" });
      }}
    >
      Preload product
    </button>
  );
};
```

This is a very contrived example. However, you could imagine doing this on whatever event: hover, focus, click or even IntersectionObserver.

<blockquote class="warning" id="preloading-hoist-static">
  **Warning:** if you have some HOC that you are using before `graphqlWithPreload`, you won't have any `preload` static property on your final component. If you want to, you will need to hoist static properties at the top level.

  ```diff
-import { withProps } from "recompose";
+import { compose, hoistStatics, withProps } from "recompose";
const EnhancedComponent = compose(
-  withRouter,
-  withProps(() => /* ... */),
+  hoistsStatics(
+    compose(
+      withRouter,
+      withProps(() => /* ... */)
+    )
+  )
  graphqlWithPreload(Query)
)(BaseComponent)
  ```
</blockquote>

## How to preload routes?

> However, in some cases you don't really know which component you should preload. This is the case for the url: `/venia-dresses` for instance. If you are familiar with Venia's sample data, you will recognize that it is a category. But you may not know which component is actually used to display it.

Reviewer: ```suggestion
However, in some cases you don't really know which component you should preload. This is the case for the url: `/venia-dresses` for instance. If you are familiar with Magento Venia's sample data, you will recognize that it is a category. But you may not know which component is actually used to display it.
```

> This is why we have implemented preloads on links too. This is possible with the React hook `theme/modules/Router/usePreload` (added in version 1.0.0-rc.0) that will give you a preload function to load the components and data for a given link.

Reviewer: ```suggestion
This is why we have implemented preloads on links too. This is possible with the React hook `theme/modules/Router/usePreload` that will give you a preload function to load the components and data for a given link.
```

For instance, let's consider that the URL `/product/:sku` should display the following component.

```js
import React from "react";
import graphqlWithPreload from "web/core/apollo/graphqlWithPreload";

const Product = props => {
  return <div>{props.product.name}</div>;
};

const EnhancedProduct = (
  graphqlWithPreload(ProductQuery, {
    options: props => ({
      variables: {
        sku: props.match.params.sku
      }
    }),
    preloadOptions: params => ({
      variables: {
        sku: params.match.params.sku
      }
    })
  })
)(Product);
```

Then this means that if you call `preload("/product/VD12")`, in `params.match.params.sku` (line 17) you will have the value `VD12`.

If you need search parameters in the URL (e.g. `?param=value`), it will be available at `params.location.search`.

## Front-Commerce Fast Mode

To conclude, if you want your application to feel even faster to your users, you need to:
* add a `preload` static property to your routes thanks to `graphqlWithPreload`
* use `usePreload` hook through the default Link component in Front-Commerce or manually in your own events.

> This also enables a wider range of optimizations such as Front-Commerce Fast Mode. Please refer to [FRONT_COMMERCE_FAST] for more information.

Reviewer: The `FRONT_COMMERCE_FAST` link is missing here.

Author: ```suggestion
This also enables a wider range of optimizations such as Front-Commerce Fast Mode. Please refer to [FRONT_COMMERCE_FAST](/docs/advanced/performance/faster-server-side-rendering.html) for more information.
```
